server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -569,6 +612,32 @@ private void allocateExistingUnassignedShards(RoutingAllocation allocation) {
         }
     }
 
+    /**
+     * Verify if all unassigned shards are allocated by the same allocator, if yes then return the allocator, else
+     * return null
+     * @param allocation {@link RoutingAllocation}
+     * @return {@link ExistingShardsAllocator} or null
+     */
+    private ExistingShardsAllocator getAndVerifySameAllocatorForAllUnassignedShards(RoutingAllocation allocation) {
```

Comment:
Why do we need this elaborate check? Simply check if there are more than one ExistingShardAllocator, bail out and use single shard allocation, which will be eventually moved to batch code path with one shardRouting.